### PR TITLE
Feature/obspace diagnostics

### PR DESCRIPTION
**Pull request by delippi**
~~Made several updates for observation space diagnostic tool. Two new files were added, "dateutils.py" and "netCDF_obspace_diag.py" for date/time manipulation as for computing the observation space statistics from netCDF diagnostic files. A function for selecting which observations to use from the diag file was also created called "filter_obs". This can filter observations by observation type from the bufr obs table (https://emc.ncep.noaa.gov/mmb/data_processing/prepbufr.doc/table_2.htm) as well as by lat/lon/pressure/error. In addition to these major changes, there were some minor updates to unit conversions the corresponding test_units.py and updated some of the example line_plot scripts (one example script was added to include more options).~~

Edit:
Added a function to help with ensemble DA diagnostic tools being developed in PyGSI. The main change includes the file "dateutils.py" which is used for date manipulation. There are various other small changes to units.py, plots.py, create_plots.py, a new example in examples, and some reorganization of tools.

### CODE REVIEW

File: src/emcpy/io/netCDF.py [hunk 1]
```diff
@@ -79,3 +92,49 @@ def read_netCDF_var(filename, variable_name, oneD=False, ftime=-1, flevel=-1):
 
     result = _np.squeeze(var)
     return result
+
+def filter_obs(code,codes,errorinv,lat,lon,pressure,\
```
Comment (CoryMartin-NOAA) [inline src/emcpy/io/netCDF.py]: This should probably be somewhere else but not immediately sure where. Any suggestions @aerorahul or @kevindougherty-noaa ?

Comment (delippi) [reply src/emcpy/io/netCDF.py]: I made a utils directory and move this function there.

File: scripts/ensemble_obspace_diag.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+# ********************************************************************
+n_mem = 30
+expt_names.append("rrfs_a_conus")
+# expt_names.append("just uncomment for a second experiment")
+
+date1 = "2022122000"
+date2 = "2022123023"  # will skip 00-18
+skip_enkf_hours = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]  # right now rrfs only runs EnKF at 18-23Z
+
+datapath = "/lfs/h2/emc/da/noscrub/donald.e.lippi/rrfs_mon/diags/"  # rrfs_a_conus/"
```
Comment (kevindougherty-noaa) [inline scripts/ensemble_obspace_diag.py]: Should we get a temp file saved locally instead of pointing to here?

File: scripts/ensemble_obspace_diag.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    # Figure
+    fig = CreateFigure(nrows=2, ncols=1, figsize=(8 * scale_fig_size, 6 * scale_fig_size))
+    fig.plot_list = [plot1, plot2]
+    fig.create_figure()  # must go before add_suptitle
+    fig.add_suptitle("Obs Space Diagnostics (%s-%s)" % (dates[0], dates[-1]), ha="center", fontsize=suptitle_fontsize)
+    fig.tight_layout()  # must go after add_suptitle
+    fig.save_figure("obs_diag_%s.png" % (ob_type))
+
+# Calculate time to run script
+utils.toc(tic, "End. ")
```
Comment (kevindougherty-noaa) [inline scripts/ensemble_obspace_diag.py]: Could we also get a `.png` of the output from this script so users can see?

File: src/emcpy/utils/dateutils.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+# times in many datasets use mixed Gregorian/Julian calendar, datetime
+# module uses a proleptic Gregorian calendar. So, I use datetime to compute
+# hours since start of Greg. calendar (15821015) and add this constant to
+# get hours since 1-Jan-0001 in the mixed Gregorian/Julian calendar.
+gregstart = datetime.datetime(1582, 10, 15)  # datetime.datetime instance
+day1 = datetime.datetime(1, 1, 1)  # datetime.datetime instance
+
+
+def dateto_hrs_since_day1CE(curdate, mixedcal=True):
+    """given datetime.datetime instance, compute hours since 1-Jan-0001"""
```
Comment (kevindougherty-noaa) [inline src/emcpy/utils/dateutils.py]: Please update docstring to Google style for `pydoc` documentation purposes. See examples here: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html as well as other functions in emcpy.

**Review by kevindougherty-noaa — CHANGES_REQUESTED** [review]
Thank you for all the additions to this repo! I had a couple small comments, with the main one being that all of your docstring in your `dateutils.py` script need to be updated to Google style so that the automated documentation works smoothly. 

I also have updated all the examples that were out of date and should be merged to `develop` shortly.

File: src/emcpy/stats/obspace.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
+import numpy as _np
```
Comment (aerorahul) [inline src/emcpy/stats/obspace.py]: @CoryMartin-NOAA 
This looks like a very specific use case.  Should this be part of `emcpy`?

File: src/emcpy/plots/create_plots.py [hunk 6]
```diff
@@ -669,26 +669,26 @@ def _set_ylim(self, ax, ylim):
         """
         ax.set_ylim(**ylim)
 
-    def _set_xticks(self, ax, xticks):
+    def _set_xticks(self, ax, xticks, latlon=False):
```
Comment (aerorahul) [inline src/emcpy/plots/create_plots.py:688]: Not sure why a generic method `_set_xticks` needs this `latlon`
Isn't the purpose of `try` already doing this?

File: scripts/ensemble_obspace_diag.py [hunk 7]
```diff
@@ -0,0 +1,320 @@
+import emcpy.utils.utils as utils
```
Comment (aerorahul) [inline scripts/ensemble_obspace_diag.py]: I don't think this belongs in `emcpy`.

File: src/emcpy/plots/create_plots.py (hunk 6 above)
Comment (delippi) [reply src/emcpy/plots/create_plots.py:688]: With the try/except method, you might axes label that look like the following (on the bottom plot). 

line_plot_options

However, I think a better solution, which I just figured out when trying to reproduce the issue, is to make sure you also use set_xticklabels like the following:
```
plot2.set_xticks(x2)
plot2.set_xticklabels([str(item) for item in x2])
```

Comment (delippi) [reply src/emcpy/plots/create_plots.py:688]: Although, I don't think the "except" ever gets invoked.

File: scripts/ensemble_obspace_diag.py (hunk 7 above)
Comment (CoryMartin-NOAA) [reply scripts/ensemble_obspace_diag.py]: I like the idea of there being some sort of 'gallery' though of useful examples for people to see how things are used, the 'examples' directory is a bit more basic whereas this would be more practical.

File: src/emcpy/plots/create_plots.py (hunk 6 above)
Comment (CoryMartin-NOAA) [reply src/emcpy/plots/create_plots.py:688]: Yeah I don't think the try, except approach is best here and like the solution of being explicit if it isn't required for the average user to do so...

File: src/emcpy/stats/obspace.py (hunk 5 above)
Comment (CoryMartin-NOAA) [reply src/emcpy/stats/obspace.py]: I guess this becomes a philosophical question of where the cutoff happens and what constitutes 'specific' vs 'generic'.

I see two approaches:
1) one 'synoptic' repository that contains generic tools and specific, shared applications of said tools
2) a generic tool repository and numerous, app/group/branch specific repositories that use the generic tools

I think both have their pros/cons. 1 runs the risk of being hard to manage, 2 runs the risk of duplicated efforts

File: scripts/ensemble_obspace_diag.py (hunk 2 above)
Comment (delippi) [reply scripts/ensemble_obspace_diag.py]: That would be nice, but I'm not certain it would make sense to do so. This script uses the RRFS ensemble diag files and there are 30 members, so for 1 day there would be about 150 files (considering just uv) since the ensemble is only run for a limited period each day. For reference, that would be 356M. I could save a subset of the ensemble though, but it is still going to be more than 1 file.

Let me know if you would still like me to add some data to point to locally.

File: scripts/ensemble_obspace_diag.py (hunk 3 above)
Comment (delippi) [reply scripts/ensemble_obspace_diag.py]: Yes, I can add this.

File: src/emcpy/utils/dateutils.py (hunk 4 above)
Comment (delippi) [reply src/emcpy/utils/dateutils.py]: I've updated the docstrings. Let me know if there is something I need to do differently. I'm still in the learning process.

**Comment by CoryMartin-NOAA** [comment]
@delippi I know @kevindougherty-noaa suggested adding test data, but these additions are like 20MB worth of diags, that is too much to carry around in the repository, because everyone would need to download that when it is cloned. Can we remove them?

File: scripts/ensemble_obspace_diag.py (hunk 7 above)
Comment (CoryMartin-NOAA) [reply scripts/ensemble_obspace_diag.py]: What about putting this script in to GSI-utils? I have priviliges there to review/merge and it might make more sense there since it is using GSI ncdiags

File: src/emcpy/stats/obspace.py (hunk 5 above)
Comment (CoryMartin-NOAA) [reply src/emcpy/stats/obspace.py]: @delippi after @aerorahul 's comment and thinking about this over the past couple weeks and now with fresh eyes, I think this functionality probably belongs in PyGSI?

Comment (delippi) [reply src/emcpy/stats/obspace.py]: @CoryMartin-NOAA I've removed this from emcpy and am working on getting it into pygsi. Thanks for the suggestion.

File: scripts/ensemble_obspace_diag.py (hunk 7 above)
Comment (delippi) [reply scripts/ensemble_obspace_diag.py]: @CoryMartin-NOAA I've moved this out of emcpy. I will work on getting it into GSI-utils as you have suggested.

File: src/emcpy/__init__.py [hunk 8]
```diff
@@ -5,8 +5,8 @@
 
 import os
 import emcpy.calcs
-import emcpy.utils
-import emcpy.stats
+import emcpy.utils.utils
```
Comment (CoryMartin-NOAA) [inline src/emcpy/__init__.py]: I don't particularly like this 'redundant' naming, can we add something to the `__init__` files so that all of the existing stuff stays `import emcpy.utils` but then there is `import emcpy.utils.dateutils` also?

File: src/emcpy/utils/dateutils.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    compute verification date given analysis date string (yyyymmddhh) and fcst hour.
+    Args:
+        analdate: (date str) analysis date
+        fcsthr: (int) forecast hour
+    Returns:
+        verifdate: (date str) verification date given analysis date string (yyyymmddhh) and fcst hour.
+    """
+    yyyy, mm, dd, hh = splitdate(analdate)
+    analdate = datetime.datetime(yyyy, mm, dd, hh)
+    verifdate = analdate + fcsthr * datetime.timedelta(hours=1)
```
Comment (CoryMartin-NOAA) [inline src/emcpy/utils/dateutils.py]: why not `hours=fcsthr`?

File: src/emcpy/utils/dateutils.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+    return a date string of the form yyyymmddhh given integers yyyy,mm,dd,hh.
+    Args:
+        yyyy: (int) year
+        mm: (int) month
+        dd: (int) day
+        hh: (int) hour
+    Returns:
+        yyyymmddhh: (date str) date
+    """
+    return "%0.4i" % (yyyy) + "%0.2i" % (mm) + "%0.2i" % (dd) + "%0.2i" % (hh)
```
Comment (CoryMartin-NOAA) [inline src/emcpy/utils/dateutils.py]: should use `f-strings` here.

File: src/emcpy/utils/dateutils.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
+# coding: utf-8 -*-
+
+"""
+utilities for working with dates using datetime module (Python 2.3 or later)
+original author: Jeff Whitaker <jeffrey.s.whitaker@noaa.gov>
+"""
+
+__all__ = [
```
Comment (CoryMartin-NOAA) [inline src/emcpy/utils/dateutils.py:8]: in general a lot of these things seem like they would be better suited creating datetime objects...

Comment (CoryMartin-NOAA) [inline src/emcpy/utils/dateutils.py:8]: in general a lot of these things seem like they would be better suited creating datetime objects...

File: src/emcpy/utils/dateutils.py [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+    Args:
+        yyyy: (int) year
+        mm: (int) month
+    Returns:
+        number of days in month given yyyy,mm
+    """
+    return calendar.monthrange(yyyy, mm)[1]
+
+
+if __name__ == "__main__":
```
Comment (CoryMartin-NOAA) [inline src/emcpy/utils/dateutils.py:231]: why is this needed?

File: src/emcpy/utils/utils.py [hunk 13]
```diff
@@ -172,3 +173,31 @@ def printcolour(text, colour='red'):
 
 
 printcolor = printcolour
+
+
+def tic():
```
Comment (CoryMartin-NOAA) [inline src/emcpy/utils/utils.py]: these function names are not exactly useful... is there a more formal timer/timing module that would be a better thing to include here? like how `logger` works for logging?

**Review by CoryMartin-NOAA — COMMENTED** [review]
sorry @delippi to keep making you change things, but I have more comments/suggestions/questions

File: src/emcpy/__init__.py (hunk 8 above)
Comment (delippi) [reply src/emcpy/__init__.py]: I think I figured out a fix.

File: src/emcpy/utils/dateutils.py (hunk 9 above)
Comment (delippi) [reply src/emcpy/utils/dateutils.py]: I updated the calculation to do
verifdate = analdate +  datetime.timedelta(hours=fcsthr)
either way should be correct though.

File: src/emcpy/utils/dateutils.py (hunk 10 above)
Comment (delippi) [reply src/emcpy/utils/dateutils.py]: I updated this to use f-strings.

File: src/emcpy/utils/utils.py (hunk 13 above)
Comment (delippi) [reply src/emcpy/utils/utils.py]: You're right about the names. These were some functions that I cooked up for myself and thought they might be useful here too. Would you rather that I remove them? I'm not sure what the "formal" method for using a timer is; I would have to do some digging. This isn't an integral part to what I need added. Let me know if I should remove or just update the naming.

File: src/emcpy/utils/dateutils.py (hunk 12 above)
Comment (delippi) [reply src/emcpy/utils/dateutils.py:231]: It would be useful for leap years and just general date calculations which might otherwise require hard coding of number of days in a month, for example.

File: src/emcpy/utils/utils.py (hunk 13 above)
Comment (CoryMartin-NOAA) [reply src/emcpy/utils/utils.py]: I suggest removing and we figure out a more robust/comprehensive timer solution later. Does that work for you?

File: src/emcpy/utils/dateutils.py (hunk 12 above)
Comment (CoryMartin-NOAA) [reply src/emcpy/utils/dateutils.py:231]: Sorry I mean the `__main__` part

File: src/emcpy/utils/dateutils.py (hunk 11 above)
Comment (delippi) [reply src/emcpy/utils/dateutils.py:8]: I think from a basic user stand point it is easier to use date strings and let these functions use datetime objects as necessary.

**Review by kevindougherty-noaa — COMMENTED** [review]
I think this all looks great. Thank you for adding the datetime applications as I think that is going to be very useful! My only request is to add a quick, basic test script and example script for using the new `horizontalspan` option.

**Review by kevindougherty-noaa — CHANGES_REQUESTED** [review]
Coming back and looking at this a second time, please make sure all docstring in your `dateutils.py` script is correct. Please use correct punctuation and include a description of what ever function does, even if the name is self explanatory. The docstring is used for when the documentation page is created, so we would like it to be as professional and clean as possible.

**Comment by delippi** [comment]
@kevindougherty-noaa, let me know if my recent changes are to your satisfaction. I added a function in "test_plots.py" to test and demo the HorizontalSpan option. I also updated the doc strings in dateutils.

**Review by kevindougherty-noaa — APPROVED** [review]
Looks great, thanks for making those changes!
